lib/V8/v8-utils.cpp

```diff
@@ -659,6 +659,105 @@ static std::string GetEndpointFromUrl(std::string const& url) {
 /// `process-utils.js` depends on simple http client error messages.
 ///   this needs to be adjusted if this is ever changed!
 ////////////////////////////////////////////////////////////////////////////////
+namespace {
+
+auto getEndpoint(v8::Isolate* isolate, std::vector<std::string>const& endpoints
+                ,std::string& url, std::string& lastEndpoint) -> std::tuple<std::string, std::string, std::string> {
+
+    // returns endpoint, relative, error
+    std::string relative;
+    std::string endpoint;
+    if (url.substr(0, 7) == "http://") {
+      endpoint = GetEndpointFromUrl(url).substr(7);
+      relative = url.substr(7 + endpoint.length());
+
+      if (relative.empty() || relative[0] != '/') {
+        relative = "/" + relative;
+      }
+      if (endpoint.find(':') == std::string::npos) {
+        endpoint.append(":80");
+      }
+      endpoint = "tcp://" + endpoint;
+    } else if (url.substr(0, 8) == "https://") {
+      endpoint = GetEndpointFromUrl(url).substr(8);
+      relative = url.substr(8 + endpoint.length());
+
+      if (relative.empty() || relative[0] != '/') {
+        relative = "/" + relative;
+      }
+      if (endpoint.find(':') == std::string::npos) {
+        endpoint.append(":443");
+      }
+      endpoint = "ssl://" + endpoint;
+    } else if (url.substr(0, 5) == "h2://") {
+      endpoint = GetEndpointFromUrl(url).substr(5);
+      relative = url.substr(5 + endpoint.length());
+
+      if (relative.empty() || relative[0] != '/') {
+        relative = "/" + relative;
+      }
+      if (endpoint.find(':') == std::string::npos) {
+        endpoint.append(":80");
+      }
+      endpoint = "tcp://" + endpoint;
+    }  else if (url.substr(0, 6) == "h2s://") {
+         endpoint = GetEndpointFromUrl(url).substr(6);
+         relative = url.substr(6 + endpoint.length());
+
+         if (relative.empty() || relative[0] != '/') {
+           relative = "/" + relative;
+         }
+         if (endpoint.find(':') == std::string::npos) {
+           endpoint.append(":80");
+         }
+         endpoint = "tcp://" + endpoint;
```

Comment:
shouldn't this be `endpoint = "ssl://" + endpoint;`?

Reply:
I am not sure the code is just moved up and it was in the original code like that. Maybe h2c does the initial connection over http. Maybe @graetzer has an opinion on this.